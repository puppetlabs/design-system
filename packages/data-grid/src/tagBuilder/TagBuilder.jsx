import React from 'react';
import { arrayOf, shape, func, string } from 'prop-types';
import { Tag, Button } from '@puppet/react-components';
import classNames from 'classnames';

import './TagBuilder.scss';

const propTypes = {
  /** Optional feature to display number of rows in table */
  filters: arrayOf(
    shape({
      /** The value returned when a remove button on a pill is clicked */
      field: string,
      /** The text shown in the first part of the pill */
      fieldLabel: string,
      /** The value shown on the second half of a pill */
      value: string,
    }),
  ).isRequired,
  /** Callback function called when user clicks a pill */
  onRemoveTag: func.isRequired,
  /** Callback function called when user clicks the remove all button */
  onRemoveAll: func.isRequired,
};

function TagBuilder({ filters, onRemoveTag, onRemoveAll }) {
  return (
    <div className="dg-tag-builder-container">
      <div className="dg-tag-container">
        {filters.map(({ field, fieldLabel, value }, idx) => {
          let displayValue = value;
          if (typeof value === 'boolean') {
            displayValue = value.toString();
          }
          const tagLabel = `${fieldLabel} = ${displayValue}`;
          return (
            <Tag
              className={classNames('dg-filter-tag', `dg-filter-tag-${idx}`)}
              onClick={() => onRemoveTag(field)}
              label={tagLabel}
              key={`dg-filter-tag-${idx + 1}`}
            />
          );
        })}
        {filters.length > 1 && (
          <Button
            className="dg-tag-remove-all-button"
            type="text"
            icon="close"
            onKeyDown={() => onRemoveAll()}
            onClick={() => onRemoveAll()}
          >
            Remove all
          </Button>
        )}
      </div>
    </div>
  );
}

TagBuilder.propTypes = propTypes;

export default TagBuilder;
