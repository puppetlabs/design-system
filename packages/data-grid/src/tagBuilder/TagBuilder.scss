@import '~@puppet/sass-variables';

.dg-tag-builder-container {
  display: flex;

  .dg-tag-container {
    display: flex;
    justify-content: space-between;
    margin-bottom: 16px;

    .rc-tag {
      margin-right: 8px !important;
    }
  }

  .dg-tag-remove-all-button {
    margin: 8px 0;
  }
}
