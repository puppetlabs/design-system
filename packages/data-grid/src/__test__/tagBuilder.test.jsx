import React from 'react';
import { mount, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import TagBuilder from '../tagBuilder/TagBuilder';

configure({ adapter: new Adapter() });

const filters = [
  {
    field: 'all-operating-system',
    fieldLabel: 'All Operating System',
    value: 'Windows',
  },
  {
    field: 'puppet-installed',
    fieldLabel: 'Puppet Installed',
    value: 'true',
  },
];

const mockfunc = jest.fn();
const mockfunc2 = jest.fn();
const wrapper = mount(
  <TagBuilder
    filters={filters}
    onRemoveTag={mockfunc}
    onRemoveAll={mockfunc2}
  />,
);

const filters2 = [
  {
    field: 'all-operating-system',
    fieldLabel: 'All Operating System',
    value: 'Windows',
  },
];

const wrapper2 = mount(
  <TagBuilder
    filters={filters2}
    onRemoveTag={mockfunc}
    onRemoveAll={mockfunc2}
  />,
);

describe('Snapshot test', () => {
  test('Check component matches previous HTML snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });
});

describe('Check component', () => {
  test('Number of tags rendered ', () => {
    expect(wrapper.find('Tag')).toHaveLength(2);
  });

  test('Remove all icon only appears when more than one tag is present ', () => {
    expect(wrapper.find('button.dg-tag-remove-all-button')).toHaveLength(1);
    expect(wrapper2.find('button.dg-tag-remove-all-button')).toHaveLength(0);
  });

  test('onRemoveTag get called when close icon is clicked ', () => {
    wrapper.find('Button.rc-tag-remove-button').first().simulate('click');
    expect(mockfunc).toHaveBeenCalled();
  });

  test('onRemoveTag get called when close icon is clicked ', () => {
    expect(mockfunc.mock.calls[0][0]).toBe('all-operating-system');
  });

  test('onRemoveAll get called when close all icon is clicked ', () => {
    wrapper.find('button.dg-tag-remove-all-button').simulate('click');
    expect(mockfunc2).toHaveBeenCalled();
  });
});
