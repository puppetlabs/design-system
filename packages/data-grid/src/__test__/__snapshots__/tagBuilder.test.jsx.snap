// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot test Check component matches previous HTML snapshot 1`] = `
<TagBuilder
  filters={
    [
      {
        "field": "all-operating-system",
        "fieldLabel": "All Operating System",
        "value": "Windows",
      },
      {
        "field": "puppet-installed",
        "fieldLabel": "Puppet Installed",
        "value": "true",
      },
    ]
  }
  onRemoveAll={[MockFunction]}
  onRemoveTag={[MockFunction]}
>
  <div
    className="dg-tag-builder-container"
  >
    <div
      className="dg-tag-container"
    >
      <Tag
        className="dg-filter-tag dg-filter-tag-0"
        emphasis="bold"
        hideRemoveButton={false}
        key="dg-filter-tag-1"
        label="All Operating System = Windows"
        onClick={[Function]}
        type="primary"
      >
        <div
          className="rc-tag rc-tag-primary rc-tag-bold dg-filter-tag dg-filter-tag-0"
        >
          <div
            className="rc-tag-label-background rc-tag-border"
          >
            <Text
              as="div"
              className="rc-tag-text"
              color={null}
              size="medium"
              style={{}}
            >
              <div
                className="rc-text rc-text-size-medium rc-tag-text"
                style={{}}
              >
                All Operating System = Windows
              </div>
            </Text>
          </div>
          <Button
            aria-label="Remove tag"
            as="button"
            className="rc-tag-remove-button"
            disabled={false}
            forwardRefAs="ref"
            icon="close"
            iconSize="small"
            innerFocus={false}
            loading={false}
            onClick={[Function]}
            trailingIcon={null}
            type="primary"
            weight="bold"
          >
            <button
              aria-label="Remove tag"
              className="rc-button rc-button-primary rc-button-bold rc-button-icon rc-button-empty rc-tag-remove-button"
              disabled={false}
              onClick={[Function]}
              type="button"
            >
              <Icon
                className="rc-button-icon-svg"
                size="small"
                style={{}}
                svg={null}
                type="close"
                viewBox={null}
              >
                <svg
                  className="rc-icon rc-icon-close rc-button-icon-svg"
                  height="12px"
                  style={
                    {
                      "height": "12px",
                      "width": "12px",
                    }
                  }
                  viewBox="0 0 8 8"
                  width="12px"
                >
                  <polygon
                    fillRule="evenodd"
                    points="6.1 .5 4 2.6 1.9 .5 .5 1.9 2.6 4 .5 6.1 1.9 7.5 4 5.4 6.1 7.5 7.5 6.1 5.4 4 7.5 1.9"
                  />
                </svg>
              </Icon>
              <span
                className="rc-button-content"
              />
            </button>
          </Button>
        </div>
      </Tag>
      <Tag
        className="dg-filter-tag dg-filter-tag-1"
        emphasis="bold"
        hideRemoveButton={false}
        key="dg-filter-tag-2"
        label="Puppet Installed = true"
        onClick={[Function]}
        type="primary"
      >
        <div
          className="rc-tag rc-tag-primary rc-tag-bold dg-filter-tag dg-filter-tag-1"
        >
          <div
            className="rc-tag-label-background rc-tag-border"
          >
            <Text
              as="div"
              className="rc-tag-text"
              color={null}
              size="medium"
              style={{}}
            >
              <div
                className="rc-text rc-text-size-medium rc-tag-text"
                style={{}}
              >
                Puppet Installed = true
              </div>
            </Text>
          </div>
          <Button
            aria-label="Remove tag"
            as="button"
            className="rc-tag-remove-button"
            disabled={false}
            forwardRefAs="ref"
            icon="close"
            iconSize="small"
            innerFocus={false}
            loading={false}
            onClick={[Function]}
            trailingIcon={null}
            type="primary"
            weight="bold"
          >
            <button
              aria-label="Remove tag"
              className="rc-button rc-button-primary rc-button-bold rc-button-icon rc-button-empty rc-tag-remove-button"
              disabled={false}
              onClick={[Function]}
              type="button"
            >
              <Icon
                className="rc-button-icon-svg"
                size="small"
                style={{}}
                svg={null}
                type="close"
                viewBox={null}
              >
                <svg
                  className="rc-icon rc-icon-close rc-button-icon-svg"
                  height="12px"
                  style={
                    {
                      "height": "12px",
                      "width": "12px",
                    }
                  }
                  viewBox="0 0 8 8"
                  width="12px"
                >
                  <polygon
                    fillRule="evenodd"
                    points="6.1 .5 4 2.6 1.9 .5 .5 1.9 2.6 4 .5 6.1 1.9 7.5 4 5.4 6.1 7.5 7.5 6.1 5.4 4 7.5 1.9"
                  />
                </svg>
              </Icon>
              <span
                className="rc-button-content"
              />
            </button>
          </Button>
        </div>
      </Tag>
      <Button
        as="button"
        className="dg-tag-remove-all-button"
        disabled={false}
        forwardRefAs="ref"
        icon="close"
        iconSize="medium"
        innerFocus={false}
        loading={false}
        onClick={[Function]}
        onKeyDown={[Function]}
        trailingIcon={null}
        type="text"
        weight="bold"
      >
        <button
          aria-label="Remove all"
          className="rc-button rc-button-text rc-button-bold rc-button-icon rc-button-full dg-tag-remove-all-button"
          disabled={false}
          onClick={[Function]}
          onKeyDown={[Function]}
          type="button"
        >
          <Icon
            className="rc-button-icon-svg"
            size="small"
            style={{}}
            svg={null}
            type="close"
            viewBox={null}
          >
            <svg
              className="rc-icon rc-icon-close rc-button-icon-svg"
              height="12px"
              style={
                {
                  "height": "12px",
                  "width": "12px",
                }
              }
              viewBox="0 0 8 8"
              width="12px"
            >
              <polygon
                fillRule="evenodd"
                points="6.1 .5 4 2.6 1.9 .5 .5 1.9 2.6 4 .5 6.1 1.9 7.5 4 5.4 6.1 7.5 7.5 6.1 5.4 4 7.5 1.9"
              />
            </svg>
          </Icon>
          <span
            className="rc-button-content"
          >
            Remove all
          </span>
        </button>
      </Button>
    </div>
  </div>
</TagBuilder>
`;
